File: extend/task_relay/hub/context_ref.py
"""HMAC signing and verification for ContextRef (M3 hardening)."""

from __future__ import annotations

import hashlib
import hmac
from typing import Any


class ContextRefSignError(Exception):
    """ContextRef signature is missing or invalid."""


def canonical_context_ref(ref: dict[str, Any]) -> bytes:
    """Build the canonical byte sequence covered by the HMAC signature."""
    uri = str(ref.get("uri") or "")
    digest = str(ref.get("sha256") or "")
    encoding = str(ref.get("content_encoding") or "")
    return f"{uri}\n{digest}\n{encoding}".encode("utf-8")


def sign_context_ref(ref: dict[str, Any], secret: str) -> str:
    """Return a hex HMAC-SHA256 signature for a ContextRef dict."""
    if not secret:
        raise ContextRefSignError("signing secret is required")
    return hmac.new(
        secret.encode("utf-8"),
        canonical_context_ref(ref),
        hashlib.sha256,
    ).hexdigest()


def verify_context_ref(ref: dict[str, Any], secret: str) -> None:
    """Raise ContextRefSignError when the ref signature is missing or invalid."""
    signature = ref.get("signature")
    if not signature:
        raise ContextRefSignError("ContextRef.signature is required")
    expected = sign_context_ref(
        {
            "uri": ref.get("uri"),
            "sha256": ref.get("sha256"),
            "content_encoding": ref.get("content_encoding") or "",
        },
        secret,
    )
    if not hmac.compare_digest(str(signature), expected):
        raise ContextRefSignError("ContextRef.signature is invalid")

File: extend/task_relay/hub/go/internal/contextref/contextref.go
package contextref

import (
	"crypto/hmac"
	"crypto/sha256"
	"encoding/hex"
	"fmt"
	"strings"
)

// Error indicates ContextRef signature is missing or invalid.
type Error struct {
	Msg string
}

func (e *Error) Error() string { return e.Msg }

// Ref carries the signed ContextRef fields.
type Ref struct {
	URI             string
	SHA256          string
	ContentEncoding string
	Signature       string
}

func canonical(ref Ref) string {
	return strings.Join([]string{ref.URI, ref.SHA256, ref.ContentEncoding}, "\n")
}

// Sign returns a hex HMAC-SHA256 signature for a ContextRef.
func Sign(ref Ref, secret string) (string, error) {
	if secret == "" {
		return "", &Error{Msg: "signing secret is required"}
	}
	mac := hmac.New(sha256.New, []byte(secret))
	_, _ = mac.Write([]byte(canonical(ref)))
	return hex.EncodeToString(mac.Sum(nil)), nil
}

// Verify raises Error when the ref signature is missing or invalid.
func Verify(ref Ref, secret string) error {
	if ref.Signature == "" {
		return &Error{Msg: "ContextRef.signature is required"}
	}
	expected, err := Sign(Ref{
		URI:             ref.URI,
		SHA256:          ref.SHA256,
		ContentEncoding: ref.ContentEncoding,
	}, secret)
	if err != nil {
		return err
	}
	if !hmac.Equal([]byte(ref.Signature), []byte(expected)) {
		return &Error{Msg: "ContextRef.signature is invalid"}
	}
	return nil
}

// RefFromMap builds a Ref from a decoded context ref dict.
func RefFromMap(raw map[string]any) (Ref, error) {
	if raw == nil {
		return Ref{}, fmt.Errorf("ref is nil")
	}
	ref := Ref{
		URI:             stringField(raw, "uri"),
		SHA256:          stringField(raw, "sha256"),
		ContentEncoding: stringField(raw, "content_encoding"),
		Signature:       stringField(raw, "signature"),
	}
	return ref, nil
}

func stringField(raw map[string]any, key string) string {
	if v, ok := raw[key].(string); ok {
		return v
	}
	return ""
}
